lib/spack/spack/binary_distribution.py

```diff
@@ -0,0 +1,111 @@
+__author__ = "Benedikt Hegner (CERN)"
+
+import llnl.util.tty as tty
+
+import os
+import platform
+import urllib2
+from architecture import get_full_system_from_platform
+from spack.util.executable import which
+import spack.cmd
+import spack
+from spack.stage import Stage
+import spack.fetch_strategy as fs
+
+def prepare():
+    """
+    Install patchelf as pre-requisite to the required relocation of binary packages
+    """
+    patchelf_spec = spack.cmd.parse_specs("patchelf", concretize=True)[0]
+    patchelf = spack.repo.get(patchelf_spec)
+    patchelf.do_install()
+
+def build_info_file(spec):
+    """
+    Filename of the binary package meta-data file
+    """
+    return os.path.join(spec.prefix,".spack","binary_distribution")
+
+def tarball_name(spec):
+    """
+    Return the name of the tarfile according to the convention
+    <architecture>-<os>-<name>-<dag_hash>.tar.gz
+    """
+    return "%s-%s-%s-%s.tar.gz" %(get_full_system_from_platform(),spec.name,spec.version,spec.dag_hash())
+
+def build_tarball(spec, outdir, force=False):
+    """
+    Build a tarball from given spec
+    """
+    tarfile = os.path.join(outdir, tarball_name(spec))
+    if os.path.exists(tarfile):
+        if force:
+            os.remove(tarfile)
+        else:
+            tty.die("file exists, use -f to force overwrite: %s"%tarfile)
+
+    tar = which('tar', required=True)
+    dirname = os.path.dirname(spec.prefix)
+    basename = os.path.basename(spec.prefix)
+    
+    # handle meta-data
+    cp = which("cp", required = True)
+    spec_file = os.path.join(spec.prefix,".spack","spec.yaml")
```

Comment:
@hegner Thanks. It works for the geant4 package I built today but not for packages that were built last. I think has to do with a change in package hashes:

 spack create_tarball  gcc@4.9.3
/usr/bin/tar: gcc-4.9.3-pewb6tme5g5rrkyqlnnirlljpzasefln: Cannot stat: No such file or directory
/usr/bin/tar: Exiting with failure status due to previous errors
==> Error: Command exited with status 2:
/usr/bin/tar --directory=/home/gartung/spack/opt/spack/linux-x86_64/gcc-4.9.3 -cvf redhat7-x86_64-gcc-pewb6tme5g5rrkyqlnnirlljpzasefln.tar.gz gcc-4.9.3-pewb6tme5g5rrkyqlnnirlljpzasefln

 spack create_tarball  geant4
geant4-10.01.p03-diz5zpnvbv3rovdetwqq2dnf4ck3qmz7/
geant4-10.01.p03-diz5zpnvbv3rovdetwqq2dnf4ck3qmz7/lib64/
geant4-10.01.p03-diz5zpnvbv3rovdetwqq2dnf4ck3qmz7/lib64/libG4graphics_reps.so
geant4-10.01.p03-diz5zpnvbv3rovdetwqq2dnf4ck3qmz7/lib64/libG4visXXX.so
geant4-10.01.p03-diz5zpnvbv3rovdetwqq2dnf4ck3qmz7/lib64/libG4track.so
....